velox/functions/prestosql/URLFunctions.h

```diff
@@ -0,0 +1,281 @@
+ * limitations under the License.
+ */
+#pragma once
+
+#include <boost/regex.hpp>
+#include <cctype>
+#include "folly/Uri.h"
+#include "velox/functions/Macros.h"
+
+namespace facebook::velox::functions {
+
+class UrlParser {
+ protected:
+  StringView submatch(const boost::cmatch& match, int idx);
+
+  template <typename TInString>
+  bool parse(TInString rawUrl, boost::cmatch& match);
+
+  bool authorityAndPathMatch(
+      boost::cmatch& urlMatch,
+      boost::cmatch& authAndPathMatch,
+      boost::cmatch& authorityMatch,
+      bool& hasAuthority);
+};
+
+FOLLY_ALWAYS_INLINE StringView
+UrlParser::submatch(const boost::cmatch& match, int idx) {
+  const auto& sub = match[idx];
+  return StringView(sub.first, sub.length());
+}
+
+template <typename TInString>
+bool UrlParser::parse(TInString rawUrl, boost::cmatch& match) {
+  static const boost::regex uriRegex(
+      "([a-zA-Z][a-zA-Z0-9+.-]*):" // scheme:
+      "([^?#]*)" // authority and path
+      "(?:\\?([^#]*))?" // ?query
+      "(?:#(.*))?"); // #fragment
+
+  if (UNLIKELY(!boost::regex_match(rawUrl.data(), match, uriRegex))) {
+    return false;
+  }
+
+  return true;
+}
+
+template <typename T>
+struct UrlExtractProtocolFunction : UrlParser {
+  VELOX_DEFINE_FUNCTION_TYPES(T);
+
+  // Results refer to strings in the first argument.
+  static constexpr int32_t reuse_strings_from_arg = 0;
+
+  // ASCII input always produces ASCII result.
+  static constexpr bool is_default_ascii_behavior = true;
+
+  FOLLY_ALWAYS_INLINE bool call(
+      out_type<Varchar>& result,
+      const arg_type<Varchar>& urlString) {
+    boost::cmatch match;
+    if (!parse(urlString, match)) {
+      result.setEmpty();
+    } else {
+      auto scheme = submatch(match, 1);
+      result.setNoCopy(scheme);
+    }
+    return true;
+  }
+};
+
+template <typename T>
+struct UrlExtractFragmentFunction : UrlParser {
+  VELOX_DEFINE_FUNCTION_TYPES(T);
+
+  // Results refer to strings in the first argument.
+  static constexpr int32_t reuse_strings_from_arg = 0;
+
+  // ASCII input always produces ASCII result.
+  static constexpr bool is_default_ascii_behavior = true;
+
+  FOLLY_ALWAYS_INLINE bool call(
+      out_type<Varchar>& result,
+      const arg_type<Varchar>& urlString) {
+    boost::cmatch match;
+    if (!parse(urlString, match)) {
+      result.setEmpty();
+    } else {
+      auto scheme = submatch(match, 4);
+      result.setNoCopy(scheme);
+    }
+    return true;
+  }
+};
+
+template <typename T>
+struct UrlExtractHostFunction : UrlParser {
+  VELOX_DEFINE_FUNCTION_TYPES(T);
+
+  // Results refer to strings in the first argument.
+  static constexpr int32_t reuse_strings_from_arg = 0;
+
+  // ASCII input always produces ASCII result.
+  static constexpr bool is_default_ascii_behavior = true;
+
+  FOLLY_ALWAYS_INLINE bool call(
+      out_type<Varchar>& result,
+      const arg_type<Varchar>& urlString) {
+    boost::cmatch match;
+    if (parse(urlString, match)) {
```

Comment:
nit: perhaps, make this check consistent with above, e.g.

```
if (!parse(urlString, match)) {
   result.setEmpty()}
} else ...
```

ditto other places